L1Trigger/L1THGCalUtilities/plugins/CaloTruthCellsProducer.cc

```diff
@@ -0,0 +1,397 @@
+};
+
+CaloTruthCellsProducer::CaloTruthCellsProducer(edm::ParameterSet const& _config) :
+  makeCellsCollection_(_config.getParameter<bool>("makeCellsCollection")),
+  caloParticlesToken_(consumes<CaloParticleCollection>(_config.getParameter<edm::InputTag>("caloParticles"))),
+  triggerCellsToken_(consumes<l1t::HGCalTriggerCellBxCollection>(_config.getParameter<edm::InputTag>("triggerCells"))),
+  // simHitsTokenEE_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsEE"))),
+  // simHitsTokenHEfront_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEfront"))),
+  // simHitsTokenHEback_(consumes<std::vector<PCaloHit>>(_config.getParameter<edm::InputTag>("simHitsHEback")))
+  dummyClustering_(_config.getParameterSet("dummyClustering"))
+{
+  produces<CaloToCellsMap>();
+  produces<l1t::HGCalClusterBxCollection>("HGCalBackendLayer1Processor2DClustering");
+  produces<l1t::HGCalMulticlusterBxCollection>("HGCalBackendLayer2Processor3DClustering");
+  if (makeCellsCollection_)
+    produces<l1t::HGCalTriggerCellBxCollection>("HGCalVFEProcessorSums");
+}
+
+CaloTruthCellsProducer::~CaloTruthCellsProducer()
+{
+}
+
+void
+CaloTruthCellsProducer::produce(edm::Event& _event, edm::EventSetup const& _setup)
+{
+  edm::Handle<CaloParticleCollection> caloParticlesHandle;
+  _event.getByToken(caloParticlesToken_, caloParticlesHandle);
+  auto& caloParticles(*caloParticlesHandle);
+
+  edm::Handle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle;
+  _event.getByToken(triggerCellsToken_, triggerCellsHandle);
+  auto& triggerCells(*triggerCellsHandle);
+
+  // std::map<uint32_t, double> hitMap(makeHitMap(_event, _setup)); // cellId -> sim energy
+  // std::map<uint32_t, double> tcMap; // tcId -> total sim energy
+
+  dummyClustering_.eventSetup(_setup);
+  showerShape_.eventSetup(_setup);
+
+  edm::ESHandle<HGCalTriggerGeometryBase> geometryHandle;
+  _setup.get<CaloGeometryRecord>().get(geometryHandle);
+  auto& geometry(*geometryHandle);
+
+  std::map<uint32_t, CaloParticleRef> tcToCalo;
+
+  // used later to order multiclusters
+  std::vector<CaloParticleRef> orderedCaloRefs;
+
+  for (unsigned iP(0); iP != caloParticles.size(); ++iP) {
+    auto& caloParticle(caloParticles.at(iP));
+    if (caloParticle.g4Tracks().at(0).eventId().event() != 0) // pileup
+      continue;
+
+    CaloParticleRef ref(caloParticlesHandle, iP);
+
+    SimClusterRefVector const& simClusters(caloParticle.simClusters());
+    for (auto& simCluster : simClusters) {
+      for (auto& hAndF : simCluster->hits_and_fractions()) {
+        DetId hitId(hAndF.first);
+        HGCalDetId tcId;
+        try {
+          tcId = HGCalDetId(geometry.getTriggerCellFromCell(hitId));
+        }
+        catch (cms::Exception const& ex) {
+          edm::LogError("CaloTruthCellsProducer") << ex.what();
+          continue;
+        }
+
+        // tcMap[tcId] += hitMap[hitId.rawId()] * hAndF.first;
+        tcToCalo.emplace(tcId, ref);
```

Comment:
The energy fractions are assigned to individual HGCal cells rather than trigger cells, so you'd need to have a scheme to split the trigger cell energy into contributions from individual constituents. But we apply some calibration on the trigger cells after we sum the energies from HGCal cells, right? So I'm not sure if this will be a well-defined procedure.

Reply:
What would be needed are the simhits. With the cell fractions + the sim cell energies, the TC fraction can be computed as the energy weighed average of the cell fractions. Then it can be applied on the trigger cell energies whatever the calibration is.
Anyway the TC calibration is done in the VFE module, so the input TCs are calibrated here.

Reply:
OK, now the trigger cell mipPt is set to
(original mipPt) * (sum simEnergy * fraction) / (sum simEnergy)
where the sum is taken over HGCal cells within the trigger cell.

This is still not totally accurate though, because I'm assuming that each trigger cell belongs to one CaloParticle. The inaccuracy will show up if we have collinear hard-scattering (or gun) particles. I don't think we'll be dealing with such cases though.

Reply:
I'm not sure that the PU particles are stored; in that case it would mean that if the fraction is not equal to 1 there are necessarily overlapping particles in bx 0.
Anyway there are overlapping particles for pi0 decays, tau decays, converted photons, showering electrons, etc.
I think we should avoid doing the assumption that each trigger cell belongs to one CaloParticle

Reply:
For studies with particle guns, pi0, tau, converted photon, etc. will all have CaloParticle of pi0, tau, and photon. CaloParticle traces the sim tracks all the way back to generator final state particles. And guns is the main use case Gianluca and I had in mind with this module.
If we want to use this with real generators that does pi0 and tau decays within, indeed there can be collinear CaloParticles with energy overlaps. Do you think we need to consider such use cases too?

Reply:
In the future certainly. Tau studies will be very important for instance.
By the way tau guns will be used and they need to have hadronic decay implemented, otherwise these samples are completely useless. Previous L1T tau gun samples didn't have the decay implemented, but I asked that the future ones have it.
For the moment it's probably ok to keep it as it is, I don't want to delay this PR more than necessary. But in the near future. more general use cases will need to be supported.

Reply:
OK, sure. The PR should be otherwise ready.